src/foam/nanos/crunch/UserCapabilityJunctionDAO.js

```diff
@@ -99,16 +104,100 @@ foam.CLASS({
       javaCode: `
       checkOwnership(x, obj);
 
+      // if the junction is being uppdated from GRANTED to EXPIRED, put into junctionDAO without checking prereqs and data
+      UserCapabilityJunction old = (UserCapabilityJunction) getDelegate().find_(x, ((UserCapabilityJunction) obj).getId());
+      if ( old.getStatus() == CapabilityJunctionStatus.GRANTED && ((UserCapabilityJunction) obj).getStatus() == CapabilityJunctionStatus.EXPIRED ) 
+        return getDelegate().put_(x, obj);
+
       boolean prereq = checkPrereqs(x, obj);
       boolean data = validateData(x, obj);
 
-      if ( prereq && data ) ((UserCapabilityJunction) obj).setStatus(CapabilityJunctionStatus.GRANTED);
+      if ( prereq && data ) {
+        ((UserCapabilityJunction) obj).setStatus(CapabilityJunctionStatus.GRANTED);
+        saveDataToDAO(x, ((UserCapabilityJunction) obj));
+        configureJunctionExpiry(x, (UserCapabilityJunction) obj, old);
+      }
       else ((UserCapabilityJunction) obj).setStatus(CapabilityJunctionStatus.PENDING);
 
       return getDelegate().put_(x, obj);
       
       `
     },
+    {
+      name: 'saveDataToDAO',
+      args: [
+        {
+          name: 'x',
+          type: 'Context'
+        },
+        {
+          name: 'obj',
+          type: 'foam.nanos.crunch.UserCapabilityJunction'
+        }
+      ],
+      documentation: `save the UserCapabilityJunction data to its dao, if daoKey is provided in the Capability model`,
+      javaCode: `
+      DAO capabilityDAO = (DAO) x.get("capabilityDAO");
+      Logger logger = (Logger) x.get("logger");
```

Comment:
Can move this above the place we use it so that we don't need to do the work of getting it out of the context when we don't need it.

Reply:
yes, moved